core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -126,25 +127,28 @@ private void addAndWaitIfFull(CompletableFuture future, List<CompletableFuture>
       }
    }
 
-   CompletableFuture<Void> handleLifespanExpireEntry(K key, V value, long lifespan) {
+   // Skip locking is available in case if the entry was found expired due to a write, which would already
+   // have the lock and remove lifespan expired is fired async
```

Comment:
I'm a bit confused... If the `RemoveExpiredCommand` is fired async, why would it need to skip locking? Also, does `available` mean `true` here?

Reply:
It needs to skip locking, because it has to remove the expired entry and generate the event before the current write can occur. The current write has the lock, so the async remove expired can't acquire locks. It waits for the remove to complete causing them to be serialized and guarantees event ordering so the expiration event comes before the write event.

Reply:
Ok, but if we block the current write until the `RemoveExpiredCommand` finishes running, I wouldn't really call it async, even if it does run on another thread.

Reply:
okay

Reply:
technically it is async from this method :)

Reply:
Sure, but that doesn't explain why it needs to skip locking :)

Reply:
Sure, it just was supposed to say required instead of available. The problem is that the write already has the lock as it mentioned. I can add another sentence stating that it could cause a deadlock.

I think the confusion is because the blocking is done by the caller. I will also add a note mentioning that when `skipLocking` is true it is up to the caller to properly ensure that the lock is held for the entire duration.

Reply:
I reworded and changed variable names so it should be more clear now.

Reply:
Thanks, it's a little more detail than I wanted now, but still much clearer :)